libraries/eosiolib/core/eosio/crypto_bls_ext.hpp

```diff
@@ -157,5 +334,259 @@ namespace eosio {
+                                                0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 
+                                                0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 
+                                                0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 
+                                                0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 
+                                                0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}; // 73x8 = 584 bytes
+    const inline std::vector<uint8_t> R1 =     {0xfd, 0xff, 0x02, 0x00, 0x00, 0x00, 0x09, 0x76, 
+                                                0x02, 0x00, 0x0c, 0xc4, 0x0b, 0x00, 0xf4, 0xeb, 
+                                                0xba, 0x58, 0xc7, 0x53, 0x57, 0x98, 0x48, 0x5f, 
+                                                0x45, 0x57, 0x52, 0x70, 0x53, 0x58, 0xce, 0x77, 
+                                                0x6d, 0xec, 0x56, 0xa2, 0x97, 0x1a, 0x07, 0x5c, 
+                                                0x93, 0xe4, 0x80, 0xfa, 0xc3, 0x5e, 0xf6, 0x15}; // 6x8 = 48 bytes
+
+    // Construct an extensible-output function based on SHA256
+    void xmd_sh256(
+        char *buf,
+        int buf_len,
+        const char *in,
+        int in_len,
+        const char *dst,
+        int dst_len
+    ) {
+        const unsigned int SHA256HashSize = 32;
+        const unsigned int SHA256_Message_Block_Size = 64;
+        const unsigned ell = (buf_len + SHA256HashSize - 1) / SHA256HashSize;
+        if (buf_len < 0 || ell > 255 || dst_len > 255) {
+            return;
+        }
+        const uint8_t Z_pad[SHA256_Message_Block_Size] = { 0, };
+        const uint8_t l_i_b_0_str[] = {
+            static_cast<uint8_t>(buf_len >> 8),
+            static_cast<uint8_t>(buf_len & 0xff),
+            0,
+            static_cast<uint8_t>(dst_len)
+        };
+        const uint8_t *dstlen_str = l_i_b_0_str + 3;
+        uint8_t b_0[SHA256HashSize];
+        bls12_381::sha256 sha;
+        sha.update(Z_pad, SHA256_Message_Block_Size);
+        sha.update(in, in_len);
+        sha.update(l_i_b_0_str, 3);
+        sha.update(dst, dst_len);
+        sha.update(dstlen_str, 1);
+        sha.digest(b_0);
+        uint8_t b_i[SHA256HashSize + 1] = { 0, };
+        for (unsigned i = 1; i <= ell; ++i) {
+            for (unsigned j = 0; j < SHA256HashSize; ++j) {
+                b_i[j] = b_0[j] ^ b_i[j];
+            }
+            b_i[SHA256HashSize] = i;
+            bls12_381::sha256 s;
+            s.update(b_i, SHA256HashSize + 1);
+            s.update(dst, dst_len);
+            s.update(dstlen_str, 1);
+            s.digest(b_i);
+            const int rem_after = buf_len - i * SHA256HashSize;
+            const int copy_len = SHA256HashSize + (rem_after < 0 ? rem_after : 0);
+            memcpy(buf + (i - 1) * SHA256HashSize, b_i, copy_len);
+        }
+    }
+    bls_s scalar_fromBE(const bls_s& in) {
```

Comment:
Should this and the methods below be inline? Otherwise user will get multi-defined symbols when this file is included by multiple files.

Reply:
that is a good suggestion! will do

Reply:
good point, let me check